src/reconcile/rec_row.c

```diff
@@ -866,23 +880,26 @@ __wt_rec_row_leaf(
                     WT_ERR(__wt_ovfl_discard_add(session, page, kpack->cell));
                 }
 
-                /*
-                 * When a tombstone without a timestamp is written to disk, remove any historical
-                 * versions that are greater in the history store for this key.
-                 */
-                if (upd_select.no_ts_tombstone && r->hs_clear_on_tombstone) {
-                    WT_ERR(__wt_row_leaf_key(session, page, rip, tmpkey, true));
-                    WT_ERR(__wt_rec_hs_clear_on_tombstone(session, r, WT_RECNO_OOB, tmpkey, false));
-                }
-
                 /* Not creating a key so we can't use last-key as a prefix for a subsequent key. */
                 lastkey->size = 0;
-
-                /* Proceed with appended key/value pairs. */
-                goto leaf_insert;
+                break;
             default:
                 WT_ERR(__wt_illegal_value(session, upd->type));
             }
+
+            /*
+             * When a tombstone without a timestamp is written to disk, remove any historical
+             * versions that are greater in the history store for this key.
+             */
+            if (upd_select.no_ts_tombstone && r->hs_clear_on_tombstone) {
+                WT_ERR(__wt_row_leaf_key(session, page, rip, tmpkey, true));
+                WT_ERR(__wt_rec_hs_clear_on_tombstone(session, r, WT_RECNO_OOB, tmpkey,
+                  upd->type == WT_UPDATE_TOMBSTONE ? false : true));
+            }
+
+            /* Proceed with appended key/value pairs. */
+            if (upd->type == WT_UPDATE_TOMBSTONE)
```

Comment:
I don't really like the refactor here. Would be better just duplicate the code in two places? @hbokhari

Reply:
We need to copy the same code into three places. Modify, standard update and tombstone. Instead bringing it down will have only one copy.